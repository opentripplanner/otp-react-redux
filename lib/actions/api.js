/* globals fetch */

import deepEqual from 'deep-equal'
import { createAction } from 'redux-actions'
import qs from 'qs'

if (typeof (fetch) === 'undefined') {
  require('isomorphic-fetch')
}

import { queryIsValid } from '../util/state'

export const receivedPlanError = createAction('PLAN_ERROR')
export const receivedPlanResponse = createAction('PLAN_RESPONSE')
export const requestPlanResponse = createAction('PLAN_REQUEST')

export function planTrip (customOtpQueryBuilder) {
  return async function (dispatch, getState) {
    const otpState = getState().otp
    const latest = otpState.searches.length && otpState.searches[otpState.searches.length - 1]
    // check for query change
    if (otpState.activeSearch !== null && latest && deepEqual(latest.query, otpState.currentQuery)) {
      console.log('query hasn\'t changed')
      return
    }
    if (!queryIsValid(otpState)) return
    dispatch(requestPlanResponse())
    const queryBuilderFn = customOtpQueryBuilder || otpState.config.customOtpQueryBuilder || constructPlanQuery
    const url = queryBuilderFn(otpState.config.api, otpState.currentQuery)
    // setURLSearch(url)
    let json
    try {
      const response = await fetch(url)
      if (response.status >= 400) {
        const error = new Error('Received error from server')
        error.response = response
        throw error
      }
      json = await response.json()
    } catch (err) {
      return dispatch(receivedPlanError(err))
    }

    dispatch(receivedPlanResponse(json))
  }
}

// function setURLSearch (params) {
//   window.location.hash = `#plan?${params.split('plan?')[1]}`
// }

function constructPlanQuery (api, query) {
  const planEndpoint = `${api.host}:${api.port}${api.path}/plan`
  const { mode, time, date } = query
  const params = {
    arriveBy: query.departArrive === 'ARRIVE',
    date,
    fromPlace: `${query.from.lat},${query.from.lon}`,
    showIntermediateStops: true,
    toPlace: `${query.to.lat},${query.to.lon}`,
    mode,
    time
  }
  const stringParams = qs.stringify(params)
  // TODO: set url hash based on params
  // setURLSearch(stringParams)
  // TODO: check that valid from/to locations are provided

  return `${planEndpoint}?${stringParams}`
}
