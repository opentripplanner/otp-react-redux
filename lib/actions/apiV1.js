import qs from 'qs'

import {
  createQueryAction,
  findGeometryForPattern,
  findGeometryForTrip,
  findPatternsForRouteError,
  findPatternsForRouteResponse,
  findRouteError,
  findRouteResponse,
  findRoutesError,
  findRoutesResponse,
  findStopsForPattern,
  findStopsForTrip,
  findStopTimesForTrip,
  findTripError,
  findTripResponse,
  receivedVehiclePositions,
  receivedVehiclePositionsError
} from './api'
import { setViewedStop } from './ui'

// Import must be done like this as maplibregl is incompatible with jest
let maplibregl = null
if (typeof jest === 'undefined') maplibregl = require('maplibre-gl')

const findTrip = (params) =>
  createQueryAction(
    `index/trips/${params.tripId}`,
    findTripResponse,
    findTripError,
    {
      noThrottle: true,
      postprocess: (payload, dispatch) => {
        dispatch(findStopsForTrip({ tripId: params.tripId }))
        dispatch(findStopTimesForTrip({ tripId: params.tripId }))
        dispatch(findGeometryForTrip({ tripId: params.tripId }))
      }
    }
  )

export function vehicleRentalQuery(
  params,
  responseAction,
  errorAction,
  options
) {
  const paramsString = qs.stringify(params)
  const endpoint = `vehicle_rental${paramsString ? `?${paramsString}` : ''}`
  return createQueryAction(endpoint, responseAction, errorAction, options)
}

/**
 * Stop times for stop query (used in stop viewer).
 */
export function findStopTimesForStop(params) {
  return function (dispatch, getState) {
    console.warn('OTP1 is not supported.')
  }
}

const getVehiclePositionsForRoute = (routeId) =>
  createQueryAction(
    `index/routes/${routeId}/vehicles`,
    receivedVehiclePositions,
    receivedVehiclePositionsError,
    {
      noThrottle: true,
      rewritePayload: (payload) => {
        return {
          routeId: routeId,
          vehicles: payload
        }
      }
    }
  )

export const findPatternsForRoute = (params) =>
  createQueryAction(
    `index/routes/${params.routeId}/patterns?includeGeometry=true`,
    findPatternsForRouteResponse,
    findPatternsForRouteError,
    {
      noThrottle: true,
      postprocess: (payload, dispatch) => {
        // load geometry for each pattern
        payload.forEach((ptn) => {
          // Some OTP instances don't support includeGeometry.
          // We need to manually fetch geometry in these cases.
          if (!ptn.geometry) {
            dispatch(
              findGeometryForPattern({
                patternId: ptn.id,
                routeId: params.routeId
              })
            )
          }
          dispatch(
            findStopsForPattern({
              patternId: ptn.id,
              routeId: params.routeId
            })
          )
        })
      },

      rewritePayload: (payload) => {
        // convert pattern array to ID-mapped object
        const patterns = {}
        payload.forEach((ptn) => {
          patterns[ptn.id] = ptn
        })

        return {
          patterns,
          routeId: params.routeId
        }
      }
    }
  )

export const findRoute = (params) =>
  createQueryAction(
    `index/routes/${params.routeId}`,
    findRouteResponse,
    findRouteError,
    {
      noThrottle: true,
      postprocess: (payload, dispatch) => {
        // load patterns
        dispatch(findPatternsForRoute({ routeId: params.routeId }))
        dispatch(setViewedStop(null))
      }
    }
  )

export function findRoutes() {
  return createQueryAction(
    'index/routes',
    findRoutesResponse,
    findRoutesError,
    {
      postprocess: (payload, dispatch) => {
        dispatch(setViewedStop(null))
      },
      rewritePayload: (payload) => {
        const routes = {}
        payload.forEach((rte) => {
          routes[rte.id] = rte
        })
        return routes
      },
      serviceId: 'routes'
    }
  )
}

export function routingQuery(searchId = null, updateSearchInReducer) {
  return function (dispatch, getState) {
    console.warn('OTP1 is not supported.')
  }
}
export default {
  findPatternsForRoute,
  findRoute,
  findRoutes,
  findStopTimesForStop,
  findTrip,
  getVehiclePositionsForRoute,
  routingQuery,
  vehicleRentalQuery
}
