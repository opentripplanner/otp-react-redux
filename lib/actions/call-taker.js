import { createAction } from 'redux-actions'
import { getUrlParams } from '@opentripplanner/core-utils/lib/query'
import { serialize } from 'object-to-formdata'
import qs from 'qs'

import { getISOLikeTimestamp } from '../util/state'
import { isModuleEnabled, Modules } from '../util/config'
import { searchToQuery, sessionIsInvalid } from '../util/call-taker'

import { resetForm } from './form'

if (typeof fetch === 'undefined') require('isomorphic-fetch')

/// PRIVATE ACTIONS

const endingCall = createAction('END_CALL')
const receivedCalls = createAction('RECEIVED_CALLS')
const receivedQueries = createAction('RECEIVED_QUERIES')
const requestingCalls = createAction('REQUESTING_CALLS')
const requestingQueries = createAction('REQUESTING_QUERIES')
const storeSession = createAction('STORE_SESSION')
const toggleFieldTrips = createAction('TOGGLE_FIELD_TRIPS')

/// PUBLIC ACTIONS

export const beginCall = createAction('BEGIN_CALL')
export const toggleCallHistory = createAction('TOGGLE_CALL_HISTORY')
export const toggleMailables = createAction('TOGGLE_MAILABLES')

/**
 * Fully reset form and toggle call history (and close field trips if open).
 */
export function resetAndToggleCallHistory() {
  return function (dispatch, getState) {
    dispatch(resetForm(true))
    dispatch(toggleCallHistory())
    if (getState().callTaker.fieldTrip.visible) dispatch(toggleFieldTrips())
  }
}

/**
 * Start a new call (and show the call history window) if
 * - the call module is enabled, and
 * - a call is not in progress, and
 * - the field trip window is not visible.
 */
export function beginCallIfNeeded() {
  return function (dispatch, getState) {
    const state = getState()
    const { activeCall, fieldTrip } = state.callTaker
    const callTakerEnabled = isModuleEnabled(state, Modules.CALL_TAKER)
    if (callTakerEnabled && !activeCall && !fieldTrip.visible) {
      dispatch(beginCall())
    }
  }
}

/**
 * Handle initializing a new Trinet session by redirecting to Trinet auth and
 * returning once authenticated successfully.
 */
function newSession(datastoreUrl, verifyLoginUrl, redirect) {
  fetch(datastoreUrl + '/auth/newSession')
    .then((res) => res.json())
    .then((data) => {
      const { sessionId: session } = data
      const windowUrl = `${verifyLoginUrl}?${qs.stringify({
        redirect,
        session
      })}`
      // Redirect to login url.
      window.location = windowUrl
    })
    .catch((error) => {
      console.error('newSession error', error)
    })
}

/**
 * Check that a particular session ID is valid and store resulting session
 * data.
 */
function checkSession(datastoreUrl, sessionId, intl) {
  return function (dispatch, getState) {
    fetch(datastoreUrl + `/auth/checkSession?sessionId=${sessionId}`)
      .then((res) => res.json())
      .then((session) => dispatch(storeSession({ session })))
      .catch((err) => {
        dispatch(storeSession({ session: null }))
        alert(
          intl.formatMessage(
            { id: 'actions.callTaker.checkSessionError' },
            { err: JSON.stringify(err) }
          )
        )
      })
  }
}

/**
 * Initialize the Call Taker and Field Trip modules by checking the session
 * query param against sessions in the datastore backend or initializing a new
 * session via Trinet.
 */
export function initializeModules(intl) {
  return function (dispatch, getState) {
    const { datastoreUrl, trinetReDirect } = getState().otp.config
    // Initialize session if datastore enabled.
    if (datastoreUrl && trinetReDirect) {
      // TODO: Generalize for non-TriNet instances.
      const sessionId = getUrlParams().sessionId
      if (sessionId) {
        // Initialize the session if found in URL query params.
        dispatch(checkSession(datastoreUrl, sessionId, intl))
      } else {
        // No sessionId was passed in, so we must request one from server.
        newSession(datastoreUrl, trinetReDirect, window.location.origin)
      }
    }
  }
}

/**
 * Fetch latest calls for a particular session.
 */
export function fetchCalls(intl) {
  return async function (dispatch, getState) {
    dispatch(requestingCalls())
    const { callTaker, otp } = getState()
    if (sessionIsInvalid(callTaker.session)) return
    const { datastoreUrl } = otp.config
    const { sessionId } = callTaker.session
    const limit = 1000
    try {
      const fetchResult = await fetch(
        `${datastoreUrl}/calltaker/call?${qs.stringify({ limit, sessionId })}`
      )
      const calls = await fetchResult.json()
      dispatch(receivedCalls({ calls }))
    } catch (err) {
      alert(
        intl.formatMessage(
          { id: 'actions.callTaker.fetchCallsError' },
          { err: JSON.stringify(err) }
        )
      )
    }
  }
}

/**
 * Store the trip queries made over the course of a call (to be called when the
 * call terminates).
 */
function saveQueriesForCall(call, intl) {
  return function (dispatch, getState) {
    const { callTaker, otp } = getState()
    const { datastoreUrl } = otp.config
    if (sessionIsInvalid(callTaker.session)) return
    if (!call) {
      alert(
        intl.formatMessage({ id: 'actions.callTaker.couldNotFindCallError' })
      )
      return
    }
    return Promise.all(
      call.searches.map((searchId) => {
        const search = otp.searches[searchId]
        const query = searchToQuery(search, call, otp.config)
        const { sessionId } = callTaker.session
        const queryData = serialize({ query, sessionId })
        return fetch(
          `${datastoreUrl}/calltaker/callQuery?sessionId=${sessionId}`,
          { body: queryData, method: 'POST' }
        )
          .then((res) => res.json())
          .catch((err) => {
            alert(
              intl.formatMessage(
                { id: 'actions.callTaker.callQuerySaveError' },
                { err: JSON.stringify(err) }
              )
            )
          })
      })
    )
  }
}

/**
 * End the active call and store the queries made during the call.
 */
export function endCall(intl) {
  return async function (dispatch, getState) {
    const { callTaker, otp } = getState()
    const { activeCall, session } = callTaker
    const { sessionId } = session
    if (sessionIsInvalid(session)) return
    // Make POST request to store new call.
    const callData = serialize({
      call: {
        endTime: getISOLikeTimestamp(otp.config.homeTimezone),
        startTime: activeCall.startTime
      },
      sessionId
    })
    try {
      const fetchResult = await fetch(
        `${otp.config.datastoreUrl}/calltaker/call`,
        { body: callData, method: 'POST' }
      )
      const id = await fetchResult.json()

      // Inject call ID into active call and save queries.
      await dispatch(saveQueriesForCall({ ...activeCall, id }, intl))
      // Wait until query was saved before re-fetching queries for this call.
      await dispatch(fetchCalls(intl))
    } catch (err) {
      console.error(err)
      alert(
        intl.formatMessage(
          { id: 'actions.callTaker.callSaveError' },
          { err: JSON.stringify(err) }
        )
      )
    }
    // Clear itineraries shown when ending call.
    dispatch(resetForm(true))
    dispatch(endingCall())
  }
}

/**
 * Fetch the trip queries that were made during a particular call.
 */
export function fetchQueries(callId, intl) {
  return function (dispatch, getState) {
    dispatch(requestingQueries())
    const { callTaker, otp } = getState()
    const { datastoreUrl } = otp.config
    if (sessionIsInvalid(callTaker.session)) return
    const { sessionId } = callTaker.session
    fetch(
      `${datastoreUrl}/calltaker/callQuery?sessionId=${sessionId}&call.id=${callId}`
    )
      .then((res) => res.json())
      .then((queries) => {
        dispatch(receivedQueries({ callId, queries }))
      })
      .catch((err) => {
        alert(
          intl.formatMessage(
            { id: 'actions.callTaker.queryFetchError' },
            { err: JSON.stringify(err) }
          )
        )
      })
  }
}
