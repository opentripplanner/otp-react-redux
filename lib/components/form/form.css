/* Plan trip button */

.otp .plan-trip-button, .view-results-button {
  width: 100%;
}
.otp .plan-trip-button[disabled] {
  opacity: 0.75;
}

.otp .bottom-fixed {
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  padding: 10px;
  height: 55px;
}

.otp .white-fade {
  background-image: linear-gradient(to bottom, rgba(255,255,255,0), rgba(255,255,255,1));
  z-index: 100;
}

/* Date-Time Modal */

.otp .date-time-modal .button-row {
  margin-bottom: 20px;
}

.otp .date-time-modal .button-row .selected {
  background-color: #ccc;
}

/* Date/Time selectors */

/* Remove up/down arrows on date/time input (to save on width) */
.otp input[type="date"].datetime-slim::-webkit-inner-spin-button {
  -webkit-appearance: none;
}

/* Prevent the calendar picker from having a white margin that covers the date */
.otp input[type="date"].datetime-slim::-webkit-calendar-picker-indicator {
  background-color: rgba(0, 0, 0, 0.0);
  margin-left: 0px;
}

.otp input[type="time"].datetime-slim::-webkit-inner-spin-button {
  -webkit-appearance: none;
}

.otp .date-time-selector .col-xs-4,
.otp .date-time-selector .col-xs-6 {
  padding-left: 5px;
  padding-right: 5px;
}

.otp .date-time-selector .date-option-button  {
  width: 100%;
  font-size: 14px;
}

.otp .date-time-selector .date-selector,
.otp .date-time-selector .time-selector {
  font-size: 16px;
  width: 100%;
  -webkit-appearance: textfield;
  -moz-appearance: textfield;
  appearance: textfield;
  box-shadow: none;
  text-align: center;
}

.otp.mobile .date-time-selector .date-selector,
.otp.mobile .date-time-selector .time-selector {
  font-size: 20px;
  padding-bottom: 10px;
}

/* switch button */
.otp .switch-button .hidden {
  display: inline-block;
  height: 0;
  overflow: hidden;
  width: 0;
}

.otp .switch-button-container {
  position: absolute;
  top: 32px;
  right: 45px;
  z-index: 100000;
}

.otp .switch-button-container-mobile {
  position: absolute;
  top: 32px;
  right: 45px;
  z-index: 100000;
}

/* Settings Preview */

.otp .settings-preview {
  cursor: pointer;
  /* Prevent text overflow from being multi-line. */
  white-space: nowrap;
}

.otp.mobile .settings-preview {
  text-align: center;
}

.otp .settings-preview .summary {
  float: left;
  min-height: 38px;
  font-size: 14px;
  line-height: 1.4;
  margin-top: -1px;
  white-space: pre-wrap;
}

.otp .settings-preview .summary.tall {
  line-height: 2.6;
}

.otp:not(.mobile) .settings-preview .summary {
  /* Prevent overflow from being multi-line. 36px is edit button width. */
  width: calc(100% - 36px);
}

.otp.mobile .settings-preview .summary {
  float: none;
  font-size: 15px;
  margin-bottom: 8px;
}

.otp .settings-preview .button-container {
  float: right;
  position: relative;
  width: 36px;
  height: 36px;
}

.otp .settings-preview .button-container .btn {
  width: 100%;
  height: 100%;
}

.otp.mobile .settings-preview .button-container {
  float: none;
  display: inline-block;
}

/* Settings Selector Panel */

.otp .settings-selector-panel .button-row {
  margin-bottom: 20px;
}

.otp .settings-selector-panel .button-row .selected {
  background-color: #ccc;
}

/* Modes Selection Panel */

.otp .modes-panel .col-xs-3,
.otp .modes-panel .col-xs-4,
.otp .modes-panel .col-xs-12 {
  padding-left: 5px;
  padding-right: 5px;
}

/* (old group-based format) */

.otp .modes-panel .mode-group-row {
  margin-bottom: 10px;
}

.otp .modes-panel .mode-group-row .group-header {
  margin-bottom: 10px;
}

.otp .modes-panel .mode-group-row .group-name {
  font-weight: 500;
  font-size: 16px;
}

.otp .modes-panel .mode-group-row .group-select {
  float: right;
  text-align: right;
  font-size: 12px;
  padding-top: 5px;
}

.otp .modes-panel .mode-group-row .mode-button-container {
  display: inline-block;
  width: 100%;
  height: 72px;
  vertical-align: top;
  position: relative;
}

.otp .modes-panel .mode-group-row .mode-button {
  border: 1px solid gray;
  padding: 3px;
  border-radius: 3px;
  width: 100%;
  height: 48px;
  text-align: center;
}

.otp .modes-panel .mode-group-row .mode-button.enabled:hover {
  background-color: #f8f8f8;
}

.otp .modes-panel .mode-group-row .mode-check {
  position: absolute;
  top: -6px;
  right: -2px;
  width: 12px;
  height: 12px;
  font-size: 16px;
}

.otp .modes-panel .mode-group-row .mode-label {
  font-size: 10px;
  text-align: center;
  line-height: 12px;
  padding-top: 4px;
}

.otp .modes-panel .mode-group-row button {
  padding: 0;
  border: none;
  background: none;
  outline: none;
}

/* General settings */

.otp.mobile .dropdown-selector {
  font-size: 16px;
  padding-top: 5px;
}

/* Error message */

.otp .error-message {
  margin: 16px;
}

.otp .error-message > .header {
  font-size: 24px;
  font-weight: 600;
  border-bottom: 1px solid #bbb;
  margin-bottom: 12px;
  padding-bottom: 2px;
}

/* Tabbed Panel */

.otp .tabbed-form-panel {
  margin-left: 10px;
  margin-right: 10px;
}

.otp .tabbed-form-panel .tab-row {
  display: table;
  width: 100%;
  table-layout: fixed;
}

.otp .tabbed-form-panel .tab {
  display: table-cell;
  width: 50%;
  vertical-align: top;
}

.otp .tabbed-form-panel .tab.left {
  padding-right: 5px;
}

.otp .tabbed-form-panel .tab.right {
  padding-left: 5px;
}

.otp .tabbed-form-panel .tab .tab-content {
  padding: 10px 15px;
  font-size: 13px;
  background: #f0f0f0;
}

.otp .tabbed-form-panel .tab.selected .tab-content {
  padding-bottom: 20px;
}

.otp .tabbed-form-panel .active-panel {
  padding: 10px 15px 5px 15px;
  background: #f0f0f0;
}

.otp .tabbed-form-panel .hide-button-row {
  text-align: center;
}

.otp .tabbed-form-panel .hide-button {
  font-size: 14px;
}

/* User settings */

.otp .user-settings {
  padding-top: 10px;
  padding-left: 8px;
  padding-right: 8px;
  padding-bottom: 20px;
  margin: 10px 10px 0 10px;
  min-height: 150px;
  background-color: #f0f0f0;
}

.otp .place-button {
  padding: 5px 0 0 0;
  text-align: left;
}

.otp .user-settings .section-header {
  font-size: small;
  color: #555;
}

.otp .place-item {
  height: 45px;
  list-style: none;
}

.otp .user-settings .place-view, .otp .user-settings .place-view {
  padding-top: 4px;
}

.otp .place-item .place-text {
  display: block;
  font-weight: 500;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.otp .place-item .place-detail {
  display: block;
  font-size: smaller;
  padding-left: 18px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.otp .user-settings hr {
  border-top: 1px solid #ccc;
  margin: 5px;
}

.otp .user-settings .btn-link {
  height: 100%;
  padding: 2px;
  border-radius: 3px;
  color: #333;
}

.otp .user-settings .btn-link:hover {
  background-color: #f5f5f5;
  border: 1px solid lightgrey;
  text-decoration: none;
}

.otp .user-settings .remember-settings .btn-link {
  margin: 0 3px;
  width: 25px;
}

.otp .user-settings .remember-settings .btn-link.active {
  background-color: #084c8d;
  color: #ffffff;
}

.otp .user-settings .remember-settings .btn-link.active:hover {
  border: 1px solid grey;
}

.otp .user-settings .btn-link:focus {
  background-color: #ddd;
  border: 1px solid grey;
  outline: none;
  text-decoration: none;
}

.otp .user-settings .disclaimer {
  font-size: x-small;
}

/* For Chrome/ium desktop browsers, if so configured,
   set the date/time input controls to automatically show the date-time picker on focus.
   This changes do not seem to affect Chrome/ium mobile browsers. */
.otp .date-time-modal .date-time-selector.with-desktop-touchscreen input {
  position: relative;
}
.otp .date-time-modal .date-time-selector.with-desktop-touchscreen input::-webkit-calendar-picker-indicator {
  background-position-x: right;
  margin-inline-end: 8px;
  position: absolute;
  right: 0;
  width: 100%;
}
.otp .date-time-modal .date-time-selector.with-desktop-touchscreen input::-webkit-datetime-edit {
  margin-inline-end: 8px;
}
