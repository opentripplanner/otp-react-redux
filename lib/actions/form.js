// TODO: Typescript
/* eslint-disable @typescript-eslint/no-use-before-define */
import { createAction } from 'redux-actions'
import coreUtils from '@opentripplanner/core-utils'
import debounce from 'lodash.debounce'
import isEqual from 'lodash.isequal'
import moment from 'moment'
import qs from 'qs'

import { queryIsValid } from '../util/state'

import {
  MobileScreens,
  routeTo,
  setMainPanelContent,
  setMobileScreen
} from './ui'
import { routingQuery } from './api'
import { setLocation } from './map'

const {
  getDefaultQuery,
  getTripOptionsFromQuery,
  getUrlParams,
  planParamsToQueryAsync
} = coreUtils.query

export const settingQueryParam = createAction('SET_QUERY_PARAM')
export const clearActiveSearch = createAction('CLEAR_ACTIVE_SEARCH')
export const setActiveSearch = createAction('SET_ACTIVE_SEARCH')
export const clearDefaultSettings = createAction('CLEAR_DEFAULT_SETTINGS')
export const storeDefaultSettings = createAction('STORE_DEFAULT_SETTINGS')

/**
 * Reset the trip form parameters to their default values.
 * @param {Boolean} [full=false] if set to true, the from/to locations and URL
 *  query parameters will also be reset.
 */
export function resetForm(full = false) {
  return function (dispatch, getState) {
    const state = getState()
    const { transitModes } = state.otp.config.modes
    if (state.otp.user.defaults) {
      dispatch(settingQueryParam(state.otp.user.defaults))
    } else {
      // Get user overrides and apply to default query
      const userOverrides = coreUtils.storage.getItem('defaultQuery', {})
      const defaultQuery = Object.assign(
        getDefaultQuery(state.otp.config),
        userOverrides
      )
      // Filter out non-options (i.e., date, places).
      const options = getTripOptionsFromQuery(defaultQuery)
      // Default mode is currently WALK,TRANSIT. We need to update this value
      // here to match the list of modes, otherwise the form will break.
      options.mode = ['WALK', ...transitModes.map((m) => m.mode)].join(',')
      dispatch(settingQueryParam(options))
    }
    if (full) {
      // If fully resetting form, also clear the active search, from/to
      // locations, and query params.
      dispatch(clearActiveSearch())
      dispatch(setLocation({ location: null, locationType: 'from' }))
      dispatch(setLocation({ location: null, locationType: 'to' }))
      // Get query params. Delete everything except sessionId.
      const params = getUrlParams()
      for (const key in params) {
        if (key !== 'sessionId') delete params[key]
      }
      dispatch(routeTo('/', qs.stringify(params, { addQueryPrefix: true })))
    }
  }
}

/**
 * Action to update any specified query parameter. Replaces previous series of
 * parameter-specific actions. If a search ID is provided, a routing query (OTP
 * search) will be kicked off immediately.
 */
export function setQueryParam(payload, searchId) {
  return function (dispatch, getState) {
    dispatch(settingQueryParam(payload))
    if (searchId) dispatch(routingQuery(searchId))
  }
}

/**
 * An action that parses the active URL's query params (or whatever is passed in)
 * and updates the state with the OTP plan query params. A new search will be
 * performed according to the behavior of setQueryParam (i.e., if the passed
 * searchId is not null).
 * @param  {Object} params an object containing URL query params
 * @param  {string} source optionally contains the source of the query params
 *                         (e.g., _CALL indicates that the source is from a past
 *                         query made during a call taker session, to prevent
 *                         a search from being added to the current call)
 */
export function parseUrlQueryString(params = getUrlParams(), source) {
  return function (dispatch, getState) {
    const state = getState()
    if (state.otp.ui.mainPanelContent !== null) return state

    // Filter out the OTP (i.e. non-UI) params and set the initial query
    const planParams = {}
    Object.keys(params).forEach((key) => {
      if (!key.startsWith('ui_')) planParams[key] = params[key]
    })
    let searchId = params.ui_activeSearch || coreUtils.storage.randId()
    if (source) searchId += source
    // Convert strings to numbers/objects and dispatch
    planParamsToQueryAsync(planParams, getState().otp.config).then((query) =>
      dispatch(setQueryParam(query, searchId))
    )
  }
}

let debouncedPlanTrip // store as variable here, so it can be reused.
let lastDebouncePlanTimeMs

/**
 * This action is dispatched when a change between the old query and new query
 * is detected. It handles checks for whether the trip should be replanned
 * (based on autoPlan strategies) as well as updating the UI state (esp. for
 * mobile).
 */
export function formChanged(oldQuery, newQuery) {
  return function (dispatch, getState) {
    const state = getState()
    const { config, currentQuery, ui } = state.otp
    const { autoPlan, debouncePlanTimeMs } = config
    const isMobile = coreUtils.ui.isMobile()
    const { fromChanged, oneLocationChanged, shouldReplanTrip, toChanged } =
      checkShouldReplanTrip(autoPlan, isMobile, oldQuery, newQuery)
    // If departArrive is set to 'NOW', update the query time to current
    if (currentQuery.departArrive === 'NOW') {
      const now = moment().format(coreUtils.time.OTP_API_TIME_FORMAT)
      dispatch(settingQueryParam({ time: now }))
    }
    // Only clear the main panel if a single location changed. This prevents
    // clearing the panel on load if the app is focused on a stop viewer but a
    // search query should also be visible.
    if (oneLocationChanged && !isMobile) {
      dispatch(setMainPanelContent(null))
    }
    if (!shouldReplanTrip) {
      // If not replanning the trip, clear the current search when either
      // location changes.
      if (fromChanged || toChanged) {
        dispatch(clearActiveSearch())
        // Return to search screen on mobile only if not currently on welcome
        // screen (otherwise when the current position is auto-set the screen
        // will change unexpectedly).
        if (ui.mobileScreen !== MobileScreens.WELCOME_SCREEN) {
          dispatch(setMobileScreen(MobileScreens.SEARCH_FORM))
        }
      }
    } else if (queryIsValid(state)) {
      // If replanning trip and query is valid,
      // check if debouncing function needs to be (re)created.
      if (!debouncedPlanTrip || lastDebouncePlanTimeMs !== debouncePlanTimeMs) {
        debouncedPlanTrip = debounce(
          () => dispatch(routingQuery()),
          debouncePlanTimeMs
        )
        lastDebouncePlanTimeMs = debouncePlanTimeMs
      }
      debouncedPlanTrip()
    }
  }
}

/**
 * Check if the trip should be replanned based on the auto plan strategy,
 * whether the mobile view is active, and the old/new queries. Response type is
 * an object containing various booleans.
 */
export function checkShouldReplanTrip(autoPlan, isMobile, oldQuery, newQuery) {
  // Determine if either from/to location has changed
  const fromChanged = !isEqual(oldQuery.from, newQuery.from)
  const toChanged = !isEqual(oldQuery.to, newQuery.to)
  const oneLocationChanged =
    (fromChanged && !toChanged) || (!fromChanged && toChanged)
  // Check whether a trip should be auto-replanned
  const strategy =
    isMobile && autoPlan?.mobile ? autoPlan?.mobile : autoPlan?.default
  const shouldReplanTrip = evaluateAutoPlanStrategy(
    strategy,
    fromChanged,
    toChanged,
    oneLocationChanged
  )
  return {
    fromChanged,
    oneLocationChanged,
    shouldReplanTrip,
    toChanged
  }
}

/**
 * Shorthand method to evaluate auto plan strategy. It is assumed that this is
 * being called within the context of the `formChanged` action, so presumably
 * some query param has already changed. If further checking of query params is
 * needed, additional strategies should be added.
 */
const evaluateAutoPlanStrategy = (
  strategy,
  fromChanged,
  toChanged,
  oneLocationChanged
) => {
  switch (strategy) {
    case 'ONE_LOCATION_CHANGED':
      if (oneLocationChanged) return true
      break
    case 'BOTH_LOCATIONS_CHANGED':
      if (fromChanged && toChanged) return true
      break
    case 'ANY':
      return true
    default:
      return false
  }
}
