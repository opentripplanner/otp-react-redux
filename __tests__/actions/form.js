/* globals describe, expect, it, jest */

import {timeoutPromise} from '../test-utils'

import {formChanged} from '../../lib/actions/form'

describe('actions > api > form', () => {
  it('should debounce numerous requests to plan a trip', async () => {
    const defaultState = {
      otp: {
        config: {
          api: {
            host: 'http://mock-host.com',
            path: '/api',
            port: 80
          },
          autoPlan: true,
          debouncePlanTimeMs: 500
        },
        currentQuery: {
          from: { lat: 12, lon: 34 },
          to: { lat: 34, lon: 12 }
        },
        searches: []
      }
    }

    const formChangedAction = formChanged()

    const mockDispatch = jest.fn()
    for (var i = 0; i < 10; i++) {
      formChangedAction(mockDispatch, () => defaultState)
    }

    // wait for request to complete
    await timeoutPromise(1000)

    expect(mockDispatch.mock.calls).toMatchSnapshot()
  })
})
