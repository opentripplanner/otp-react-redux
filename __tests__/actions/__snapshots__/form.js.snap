// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions > api > form should debounce numerous requests to plan a trip 1`] = `
Array [
  Array [
    Object {
      "type": "FORM_CHANGED",
    },
  ],
  Array [
    Object {
      "type": "FORM_CHANGED",
    },
  ],
  Array [
    Object {
      "type": "FORM_CHANGED",
    },
  ],
  Array [
    Object {
      "type": "FORM_CHANGED",
    },
  ],
  Array [
    Object {
      "type": "FORM_CHANGED",
    },
  ],
  Array [
    Object {
      "type": "FORM_CHANGED",
    },
  ],
  Array [
    Object {
      "type": "FORM_CHANGED",
    },
  ],
  Array [
    Object {
      "type": "FORM_CHANGED",
    },
  ],
  Array [
    Object {
      "type": "FORM_CHANGED",
    },
  ],
  Array [
    Object {
      "type": "FORM_CHANGED",
    },
  ],
  Array [
    [Function],
  ],
]
`;
